### ISSUE TITLE
glmer offset problems

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#273]
Two similar problems reported by Steve Bellan and Xavier Paoletti.

Probably some sort of internal overflow (verbose=100 shows that we are getting `nan` values), but it's hard to see immediately where they're coming from since I thought we clamped all the relevant derivative and inverse-link functions to avoid this sort of problem ...

```
library("lme4")
load(system.file("testdata","sbTobb.Rdata",package="lme4"))
glmer(infected ~ vacc + (1|cluster) + offset(log(perstime)), 
      family=binomial(link='cloglog'),  data = temp,
      verbose=100)

dd <- read.table(system.file("testdata",
                             "dataEx_Glmer.txt",package="lme4"),
                 header=TRUE)
slope <- 1.5
glmer(DLT ~ offset(slope*dose) + (1 | patid),
      family=binomial,data=dd)
```

**Comment by stevencarlislewalker** [comment]
As usual I could be off here, but it looks like `clogloginv` is only bounded below one, not above zero.
```{c}
    template<typename T>
    struct clogloginv : public std::unary_function<T, T> {
	const T operator() (const T& x) const {
	    return T(std::min(1.-std::numeric_limits<T>::epsilon(),
                              pgumbel2(double(x), 0., 1., 1)));
	}
    };
```
Contrast this with `R`'s `clogloginv`,
```{r}
pmax(pmin(-expm1(-exp(eta)), 1 - .Machine$double.eps), .Machine$double.eps)
```

**Comment by bbolker** [comment]
I added the lower bound and got somewhat farther:

```
*** pwrssUpdate step 0
pdev=8.27503e+09; delu_min: -28564.7; delu_max: 28308.6; delb_min: -54659; delb_max: -28089.5
*** pwrssUpdate step 1
pdev=916549; delu_min: -439.197; delu_max: 47.3528; delb_min: -1.09888e+14; delb_max: 1.73208e+14
*** pwrssUpdate step 2
pdev=716149; delu_min: -15.7878; delu_max: 35.8027; delb_min: -4.38236e+15; delb_max: 5.40432e+12
*** pwrssUpdate step 3
pdev=32184.8; delu_min: -15.7878; delu_max: 35.8027; delb_min: -4.3829e+15; delb_max: 6.92645e+13
*** pwrssUpdate step 4
pdev=32184.8; delu_min: -15.7878; delu_max: 35.8027; delb_min: -4.26761e+15; delb_max: 1.72926e+13
npt = 3 , n =  1 
rhobeg =  0.2 , rhoend =  2e-07 
start par. =  1 fn =  32184.82 
  1:     32184.816:  1.00000
  2:     33880.798:  1.20000
  3:     30797.194: 0.800000
  4:     29717.933: 0.600000
  5:     28484.492: 0.200000
  6:     28330.312: 9.69225e-14
rho:    0.020 eval:   6 fn:      28330.3 par:9.69225e-14 
  7:     28331.853: 0.0200000
rho:   0.0020 eval:   7 fn:      28330.3 par:9.69225e-14 
  8:     28330.327: 0.00200000
rho:  0.00020 eval:   8 fn:      28330.3 par:9.69225e-14 
  9:     28330.312: 0.000200000
rho:  2.0e-05 eval:   9 fn:      28330.3 par:9.69225e-14 
 10:     28330.312: 2.00000e-05
 11:     28330.312: 1.00000e-05
rho:  2.0e-06 eval:  11 fn:      28330.3 par:9.69225e-14 
 12:     28330.312: 2.00000e-06
rho:  2.0e-07 eval:  12 fn:      28330.3 par:9.69225e-14 
 13:     28330.312: 2.00000e-07
 14:     28330.312: 1.00000e-07
 15:     28330.312:  0.00000
At return
eval:  15 fn:      28330.312 par: 9.69225e-14
Error: cannot generate feasible simplex
```

**Comment by bbolker** [comment]
... and with `control=glmerControl(optimizer="bobyqa")` I actually get a result, but not a sensible one ...

**Comment by stevencarlislewalker** [comment]
And now I can generate a deviance function, which before was not possible.

**Comment by bbolker** [comment]
PS: a similar workaround to the one suggested in Paoletti's case, i.e. shifting the offset, seems to work:
```
glmer(infected ~ vacc + (1|cluster) + offset(I(log(perstime)-5)), 
        family=binomial(link='cloglog'),  data = temp, control=glmerControl(optimizer="bobyqa"))
```
seems to work just fine (it seems to work OK with a shift as small as 2, but not with shift=1).  This should make it possible to get answers (all that needs to be done *post hoc* is to shift the estimated intercept by an appropriate amount), so I now consider this problem minimally "worked around", although I still think it constitutes a bug/unwarranted fragility.

**Comment by stevencarlislewalker** [comment]
This code suggests to me that the optimum is on the boundary.  What do you think?
```{r}
parsedForm <- glFormula(infected ~ vacc + (1|cluster) + offset(log(perstime)), 
                        family=binomial(link='cloglog'),  data = temp,
                        verbose=100)
dfun <- do.call(mkGlmerDevfun, parsedForm)
th <- seq(0.01, 1, length = 100)
dv <- sapply(th, dfun)
plot(th, dv, type = "l")
```

**Comment by bbolker** [comment]
I think the estimated variance (when you adjust the offset shift to make things work) is *not* 0 (it's around 0.66), which puzzles me: the only thing I can think of (haven't followed up yet) is that this is a place where the difference between the two stages of GLMM computation (nAGQ=0 vs nAGQ=1) might be relevant?

**Comment by stevencarlislewalker** [comment]
You are almost certainly correct about `nAGQ` @bbolker, and this reveals something more interesting. I think this is an example that the initial `nAGQ = 0` steps were the cause of the problem.  Maybe the accuracy of the `nAGQ = 0` approximation is just too crude when probabilities get as close as they do here to the boundaries.  Stepping through `debug(glmer)` while fitting to the original problem, the first `nAGQ = 0` step pushes `theta` down to `1.42e-12`, and the `nAGQ = 1` step is apparently not able to get to a good solution from there.  Also, we don't need to shift the offset if we manually skip the `nAGQ = 0` step:
```{r}
parsedForm <- glFormula(infected ~ vacc + (1|cluster) + offset(log(perstime)), 
                        family=binomial(link='cloglog'),  data = temp)
dfun <- do.call(mkGlmerDevfun, parsedForm)
dfun <- updateGlmerDevfun(dfun, parsedForm$reTrms)
optim(c(1, -8, 0), dfun, lower = c(0, -Inf, -Inf),
          method = "L-BFGS-B")
## $par
## [1]  0.6617482 -8.3594946 -0.7835467
## 
## $value
## [1] 4416.466
## 
## $counts
## function gradient 
##       18       18 
## 
## $convergence
## [1] 0
```

**Comment by bbolker** [comment]
... although note that starting from a low enough intercept value is also necessary.  (1, 0, 0) fails, (1,-4,0) works.  

```
library("optimx")
 optimx(c(1,-4,0), dfun2, lower = c(0, -Inf, -Inf),
           method = c("L-BFGS-B", "bobyqa", "nlminb"))
```
gives

```
              p1     p2      p3 value fevals gevals niter    xtimes
L-BFGS-B  0.6596 -8.358 -0.7828  4416     29     29    NA     59.41
 bobyqa   0.6715 -8.386 -0.7689  4416    331     NA    NA    100.52
nlminb    0.6616 -8.359 -0.7833  4416     24     75    19     25.41
```

(with `convcode=0`, `kkt1=kkt2=TRUE` for all three optimizers).  `bobyqa` might be so much slower because of its default strict tolerances (I'm pretty sure this is the `minqa` implementation), although the more precise NLL for L-BFGS-B and nlminb is 4416.466 vs. 4416.500 for bobyqa ...

**Comment by stevencarlislewalker** [comment]
You're right again @bbolker.  I forgot to reset to `1, 0, 0` and got overly excited.

**Comment by bbolker** [comment]
it does suggest that if there were a way to tell `glmer` to bypass `nAGQ=0` (I don't think there is), it would make a `glmer`-based (no need to mess around with `?modular` stuff) workaround possible by turning off `nAGQ=0` and using appropriate starting values.  It would be pretty easy to implement; I'm just worried about argument-list-length creep.

**Comment by stevencarlislewalker** [comment]
I was thinking the same thing.  Then this problem would be solved by shutting off `nAGQ = 0` and changing starting values -- no need for the offset shifting.  And I don't think we need to worry about argument creep.

**Comment by stevencarlislewalker** [comment]
With 1f66565 and 37418fb there is now a clear way to fit this model by setting better initial values and skipping the initial `nAGQ = 0` step:
```{r}
glmer(infected ~ vacc + (1|cluster) + offset(log(perstime)), 
      family=binomial(link='cloglog'),  data = temp,
      start = list(fixef = c(-4, 0)),
      control = glmerControl(nAGQ0initStep = FALSE))
```

**Comment by bbolker** [comment]
Unfortunately, the other example isn't working yet

```
dd <- read.table(system.file("testdata",
                             "dataEx_Glmer.txt",package="lme4"),
                 header=TRUE)
slope <- 1.5
glmer(DLT ~ offset(slope*dose) + (1 | patid),
      family=binomial,data=dd,
      start=list(theta=1,beta=-6),
      control=glmerControl(nAGQ0initStep = FALSE,
                           optimizer="bobyqa"),
      verbose=100)
```

I added some additional clamping to `logitinv`, and enabled debugging in `external.cpp::internal_glmerWrkIter`, which produces:

```
[first step]
(igWI, pre-updateXwts) Xwts: min: 1 sqrtWrkWt: min: 0.00641352
(igWI) Xwts: min: 0.00641352 max: 0.079494
(igWI) delu_min: -2.93564; delu_max: 26.2819; delb_min: -877.785; delb_max: -877.785
(igWI) mu: min: 2.22045e-16 max: 2.22045e-16
[clearly X-weights have already been clamped]

[second step]
(igWI, pre-updateXwts) Xwts: min: 0.00641352 sqrtWrkWt: min: 1.49012e-08
(igWI) Xwts: min: 1.49012e-08 max: 1.49012e-08
(igWI) delu_min: -1; delu_max: 2; delb_min: 7.506e+14; delb_max: 7.506e+14
(igWI) mu: min: 1 max: 1

[PWRSS loop]
pwrssUpdate: Entering step halving loop
step-halving iteration 0:  pdev=10994.8; delu_min: -1.96782; delu_max: 12.641; delb_min: 3.753e+14; delb_max: 3.753e+14
step-halving iteration 1:  pdev=11239.3; delu_min: -2.45173; delu_max: 19.4614; delb_min: 1.8765e+14; delb_max: 1.8765e+14
step-halving iteration 2:  pdev=11403.6; delu_min: -2.69368; delu_max: 22.8717; delb_min: 9.3825e+13; delb_max: 9.3825e+13
step-halving iteration 3:  pdev=11496.2; delu_min: -2.81466; delu_max: 24.5768; delb_min: 4.69125e+13; delb_max: 4.69125e+13
step-halving iteration 4:  pdev=11545.1; delu_min: -2.87515; delu_max: 25.4293; delb_min: 2.34562e+13; delb_max: 2.34562e+13
step-halving iteration 5:  pdev=11570.2; delu_min: -2.90539; delu_max: 25.8556; delb_min: 1.17281e+13; delb_max: 1.17281e+13
step-halving iteration 6:  pdev=11582.9; delu_min: -2.92051; delu_max: 26.0688; delb_min: 5.86406e+12; delb_max: 5.86406e+12
step-halving iteration 7:  pdev=11589.3; delu_min: -2.92807; delu_max: 26.1753; delb_min: 2.93203e+12; delb_max: 2.93203e+12
step-halving iteration 8:  pdev=11592.6; delu_min: -2.93185; delu_max: 26.2286; delb_min: 1.46602e+12; delb_max: 1.46602e+12
step-halving iteration 9:  pdev=11594.2; delu_min: -2.93375; delu_max: 26.2553; delb_min: 7.33008e+11; delb_max: 7.33008e+11
Error: (maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate
```

**Comment by stevencarlislewalker** [comment]
I have a suspicion that this is still related to the `nAGQ = 0` problem.  The difference is that it in this case it is not even possible to construct an `nAGQ = 0` deviance function, let alone optimize it.

**Comment by bbolker** [comment]
I added a lower bound on `logitinv` (just pushed it now), which was also missing.  See if that lets you construct the deviance function ...

**Comment by stevencarlislewalker** [comment]
I think my suspicion is correct.  I just made a change to keep `nAGQ = 0` deviance functions from being constructed when `glmerControl(nAGQ0initStep = FALSE)`, and it leads to the following.  Just running `R CMD check` now.
```{r}
> glmer(form, family = binomial, data = dd,
+       start = list(fixef = 0),
+       control = glmerControl(
+           nAGQ0initStep = FALSE,
+           optimizer = "bobyqa"))
Generalized linear mixed model fit by maximum likelihood (Laplace
  Approximation) [glmerMod]
 Family: binomial  ( logit )
Formula: DLT ~ offset(slope * dose) + (1 | patid)
   Data: dd
     AIC      BIC   logLik deviance df.resid 
154.0657 160.4517 -75.0329 150.0657      178 
Random effects:
 Groups Name        Std.Dev. 
 patid  (Intercept) 1.745e-08
Number of obs: 180, groups:  patid, 30
Fixed Effects:
(Intercept)  
     -9.828  
```

**Comment by stevencarlislewalker** [comment]
ca3125302af seems to have fixed this issue.  I'll leave it open until tests have been written.

**Comment by bbolker** [comment]
Thanks.  It would still be nice (wishlist) if these examples Just Worked, but that seems much lower priority than having an easy & straightforward workaround implemented ...

**Comment by stevencarlislewalker** [comment]
@bbolker, one simple approach to Just Working would be to wrap `try`-type commands around the initial `nAGQ = 0` step, and automatically skip it if there are errors.

**Comment by bbolker** [comment]
yes, but starting values might also need to be tweaked (and `qlogis(mean(response))` wouldn't work, need to take offsets into account)

**Comment by mmaechler** [comment]
I notice that we do not store the `glmerControl` structure that was in place during the computation.
I think we should store it with the object...in that case, the object could also store the information that
the  'nAGQ=0' gave an error and was skipped..

**Comment by stevencarlislewalker** [comment]
So if we started fixed intercepts at `qlogis(mean(response))` as well as skip `nAGQ = 0`, do you think that  that would have a chance of working @bbolker?

That's a good point @mmaechler.  If this `try` idea is implemented, I agree that information about the first fitting attempt should be stored in the object.

**Comment by stevencarlislewalker** [comment]
...of course starting at `qlogis(mean(response))` should account for any `offset`s (and `weights`?).

**Comment by dmbates** [comment]
The example seems so weird that I am not surprised that it is difficult to fit the model.  Using very large offset values like that is bound to lead to trouble.  And the final fit gives a random-effects variance that is essentially zero so it ends up being a one-dimensional optimization.  By the way, I imagine the Laplace approximation works with a variance of zero but does the Gauss-Hermite quadrature code give a sensible answer in that case.  I probably didn't consider that possibility in the original code.

This is all related to coming up with better starting estimates for the parameters.  It reminds me of the NIST examples for nonlinear regression (the NISTnls package) where a "difficult" example consists of taking a straightforward problem and using ridiculous starting values.

The reason that I wanted the nAGQ=0 fit was to refine the starting estimates for both the fixed-effects and the covariance parameters before starting either Laplace or Gauss-Hermite iterations.

In the larger view I get concerned about adding so many options to handle corner cases.  It complicates the code and makes it difficult to test the many possible combinations.  As in the NISTnls examples I don't feel it is the responsibility of those writing the optimization method to handle all possible cases, no matter how misguided they may be.

Perhaps a better approach is to think of more creative ways of devising or refining starting estimates.

**Comment by stevencarlislewalker** [comment]
I take your point about testing code with too many options.  Similarly, I suspect that including mechanisms in `glmer` for constructing more creative starting estimates could be a bad idea from the perspective of code maintenance.    However, perhaps a user-callable function for suggesting starting estimates could be worthwhile?  This would force users to think a little more about their difficult model fitting problems.

Regarding the case of `nAGQ > 1`, it seems that it still gives sensible answers when the random effects variance is close to zero (not shown). 

I also take your point about code maintenance and testing in the specific instance of adding a control parameter to turn off the initial `nAGQ = 0` fit.  This is why I suggested separating this fitting strategy from the previous strategy as much as possible (#276).  However, given that in some circumstances `nAGQ = 0` can effectively lead to poor starting estimates, it seems like having a mechanism for turning it off is a good idea.  At the very least, it doesn't seem to mess up very easy examples (although there are differences):
```{r}
> nAGQ <- c(1, 5, 10, 20)
> nAGQ0 <- c(FALSE, TRUE)
> cntrl <- expand.grid(nAGQ = nAGQ, nAGQ0 = nAGQ0)
> estPars <- list()
> for(i in 1:nrow(cntrl)) {
+     m <- glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
+                family = binomial, data = cbpp, nAGQ = cntrl[i, "nAGQ"],
+                control = glmerControl(nAGQ0initStep = cntrl[i, "nAGQ0"]))
+     estPars[[i]] <- unlist(getME(m, c("theta", "beta")))
+ }
> cbind(cntrl, t(simplify2array(estPars)))
  nAGQ nAGQ0 theta.herd.(Intercept)     beta1      beta2     beta3     beta4
1    1 FALSE              0.6422551 -1.398531 -0.9923310 -1.128676 -1.580327
2    5 FALSE              0.6473518 -1.399186 -0.9914359 -1.127846 -1.579520
3   10 FALSE              0.6475148 -1.399228 -0.9914072 -1.127832 -1.579467
4   20 FALSE              0.6475132 -1.399225 -0.9914083 -1.127834 -1.579469
5    1  TRUE              0.6420699 -1.398343 -0.9919250 -1.128216 -1.579745
6    5  TRUE              0.6473692 -1.399202 -0.9914380 -1.127859 -1.579506
7   10  TRUE              0.6475199 -1.399224 -0.9914089 -1.127810 -1.579481
8   20  TRUE              0.6475199 -1.399224 -0.9914089 -1.127810 -1.579481
```
